### ISSUE TITLE
Format day of month ('Do') french translation error

### ISSUE THREAD

**Issue by peterpeterparker** [date-fns/date-fns#437]
Hi,

I think that the format of the day of month ('Do') isn't correct in french, it doesn't follow the same rules as in english or german for example or even doesn't follow the same logic as day of the year of week in french.

day of month ('Do'):

1st, 2nd, ..., 31st in english is correct
or
1., 2., ... 31. in german is correct too

but for day of the month in french, you not gonna says
1er, 2e, ... 31e 

but 

1er, 2, ... 31 

like "le 1er janvier 2017" but like "le 2 janvier 2017", ... "le 31 janvier 2017"

Therefore I think that the function ordinal in src/locale/fr/build_format_locale/index.js should be extended, when the number is greater than 1, the code "number + 'e'" should only apply for day of the week and day of the year where it should only return "number" when day of the month.

Best regards

**Comment by kossnocorp** [comment]
Thank you for the report! cc @date-fns/i18n-french

**Comment by fbonzon** [comment]
I'm trying to fix the exact same problem in Moment.js. (And might fix it here too.) Please read related issue https://github.com/moment/moment/issues/3375#issuecomment-280949260

The `Do` token is correctly implemented here, in my understanding. It's a question of definition and documentation. The definition of `Do` is "day of month with ordinal suffix". Example usage:

#### English
2017-03-01 → `Do` day of the month → 1st day of the month
2017-03-23 → `Do` day of the month → 23rd day of the month
2017-03-24 → `Do` day of the month → 24th day of the month

#### French
2017-03-01 → `Do` jour du mois → 1er jour du mois
2017-03-23 → `Do` jour du mois → 23e jour du mois
2017-03-24 → `Do` jour du mois → 24e jour du mois

When writing long date format in French, we want to use ordinal for day 1 (token `Do`) and cardinal for other days (token `D`). There is currently no way to express this in the locale files.

Solutions I can think of:

1. Create a new token with definition "day of month for long date format (with cardinal or ordinal depending on locale)". For French, it would return the same value as `Do` for day 1, and same value as `D` otherwise.

Name it `Dom` for Day of month (latest suggestion in the above referred issue). Or `F` (next letter after D, and E is already taken) ?

Or maybe this problem is too specific of the French language, doesn't merit a dedicated token for all languages, and is semantically better handled within the French locale code. This brings second solution:

- Implement the equivalent of `L`, `LL`, `LLL`, `LLLL` formats in Moment.js. Make the format depend on the date, and return in French, e.g. for `LL`:
`Do MMMM YYYY` when day = 1
`D MMMM YYYY` otherwise

Other ideas? What is your preference?

**Comment by leshakoss** [comment]
```javascript
var monthsTokens = ['MMM', 'MMMM']
monthsTokens.forEach(function (monthToken) {
  formatters['Do ' + monthToken] = function (date) {
    var dayOfMonth = date.getUTCDate()

    var formattedDayOfMonth = dayOfMonth === 1 ? '1er' : String(dayOfMonth)
    var formattedMonth = formatters[monthToken](date)

    return formattedDayOfMonth + ' ' + formattedMonth
  }
})
```

Maybe something like that? (haven't tested it) It makes formatters 'Do MMM' and 'Do MMMM'. We did the similar thing in Russian locale: https://github.com/date-fns/date-fns/blob/master/src/locale/ru/build_format_locale/index.js#L75

**Comment by leshakoss** [comment]
@fbonzon oh yeah, I missed your pull request, sorry 😅

**Comment by peterpeterparker** [comment]
I'm agree with your comment @fbonzon 

Maybe other languages, like Italian, follow the same rules? Like "Il 1° guigno" ("the 1th june") but "Il 2 guigno" ("the 2nd june").

About your ideas of solution 1. or 2., I personally think both achieve the goals and can't tell wich one is the better ;)

**Comment by fbonzon** [comment]
@leshakoss #448 addresses an unrelated problem, thus I did not mention it here.

A similar code to the Russian example you mentioned could be done for French, yes.

If I understand correctly, it is possible to include a given token in several formatters? How is selected the one to use? E.g. if I define formatters for
`MMM`
`Do MMM`
How is `MMM` formatted? With the second one because is longer token? Or the order of formatters definitions is relevant, and the first (or last one) matching is selected?

**Comment by leshakoss** [comment]
@fbozon both formatters (MMM and Do MMM) should work, I think

**Comment by leshakoss** [comment]
@fbonzon both formatters (MMM and Do MMM) should work, I think

**Comment by fbonzon** [comment]
Yes, but what is the logic to decide which formatter to use? I could dig into the code, but maybe you know the answer already.

Provided I define formatters for (1) `MMM`, (2) `Do` and (3) `Do MMM`, there is an ambiguity for the formatter to use to display `Do` and `MMM` when they appear in format strings. Which formatter has priority?

**Comment by leshakoss** [comment]
@fbonzon alright

token string is splitted by a regexp of all possible tokens, sorted backward alphanumerically:
https://github.com/date-fns/date-fns/blob/master/src/locale/_lib/build_formatting_tokens_reg_exp/index.js#L20
So, regexp would prefer first `^MMM`, then `^Do MMM`, then `^Do`. At least, that's how it works in Russian locale:

```
> format(new Date(), 'MMMM Do Do MMMM Do', {locale: ruLocale})
'февраль 23-е 23-е февраля 23-е'
```
Here, it splitted the token string like this: ['MMMM', ' ', 'Do', ' ', 'Do MMMM', ' ', 'Do']

**Comment by fbonzon** [comment]
Is now clear to me, thanks. Verified in code, is the same for all locales. The priority is indeed defined only by reversed alphabetical order.

It would work for our use case. We would modify contents of `Do MMM` and `Do MMMM`, by omitting the ordinal suffix for days other than 1.

I can code this is it's your preferred solution. It's not the most elegant one, in my opinion, but it works.

**Comment by leshakoss** [comment]
@fbonzon yeah alright thank you 😄

**Comment by fbonzon** [comment]
Fix released in v1.28.0

**Comment by antoinerousseau** [comment]
Could we get a similar fix in v2?
